// @refinedev/inferencer/dist/headless.js
var Xt=Object.create;var be=Object.defineProperty;var er=Object.getOwnPropertyDescriptor;var tr=Object.getOwnPropertyNames;var rr=Object.getPrototypeOf,nr=Object.prototype.hasOwnProperty;var o=(e,t)=>be(e,"name",{value:t,configurable:!0});var or=(e,t)=>{for(var c in t)be(e,c,{get:t[c],enumerable:!0})},ut=(e,t,c,a)=>{if(t&&typeof t=="object"||typeof t=="function")for(let s of tr(t))!nr.call(e,s)&&s!==c&&be(e,s,{get:()=>t[s],enumerable:!(a=er(t,s))||a.enumerable});return e};var K=(e,t,c)=>(c=e!=null?Xt(rr(e)):{},ut(t||!e||!e.__esModule?be(c,"default",{value:e,enumerable:!0}):c,e)),sr=e=>ut(be({},"__esModule",{value:!0}),e);var Cr={};or(Cr,{HeadlessCreateInferencer:()=>Me,HeadlessCreateRenderer:()=>ot,HeadlessEditInferencer:()=>Oe,HeadlessEditRenderer:()=>st,HeadlessInferencer:()=>kr,HeadlessListInferencer:()=>Ee,HeadlessListRenderer:()=>nt,HeadlessShowInferencer:()=>Ve,HeadlessShowRenderer:()=>rt});module.exports=sr(Cr);var Ae=K(require("react")),Zt=require("@refinedev/core");var G=K(require("react")),$e=require("@refinedev/core");var ar=/(image|photo|avatar|url|logo|cover|thumbnail|icon)/i,De=o((e,t)=>{let c=["name","label","title","count","content","username","nickname","login","firstName","lastName","url"];ar.test(e)&&c.unshift("url");let a=c.find(s=>s in t&&!!t[s]);if(a)return a==="firstName"&&"lastName"in t?["firstName","lastName"]:a},"getFieldableKeys");var lt=require("@refinedev/core");function He(e,t,c){if(t)return t;let a=c==null?void 0:c.find(p=>p.name===e),s=(0,lt.pickNotDeprecated)(a==null?void 0:a.meta,a==null?void 0:a.options);return s!=null&&s.dataProviderName?s.dataProviderName:"default"}o(He,"pickDataProvider");var ke=o(e=>{var t;return(t=e==null?void 0:e.options)==null?void 0:t.dataProviderName},"dataProviderFromResource");var je=o(e=>{let t=[...e].filter(Boolean).sort((c,a)=>(a.priority||0)-(c.priority||0))[0];return t||null},"pickInferredField");var mt=K(require("prettier/standalone")),dt=K(require("prettier/parser-typescript"));var Be=o(e=>{try{return mt.default.format(e,{parser:"typescript",plugins:[dt.default],arrowParens:"always",trailingComma:"all",semi:!0,tabWidth:4,printWidth:80})}catch(t){return console.warn("Formatting error in inferencer",t),e}},"prettierFormat");var ft=o(e=>e.replace(/ {2,}/g," ").replace(/\n{2,}/g,`
`),"prettySpaces");var ze=/(-id|-ids|_id|_ids|Id|Ids|ID|IDs)(\[\])?$/,gt=o((e,t)=>{let c=ze.test(e),a=typeof t=="string"||typeof t=="number",s=Array.isArray(t)&&t.every(p=>typeof p=="string"||typeof p=="number");return c&&a||c&&s?{key:e,relation:!0,multiple:!!s,type:"relation",priority:1}:!1},"relationInfer");var z=o(e=>e.replace(ze,""),"removeRelationSuffix");var U=o(e=>z(e).replace(/([a-z])([A-Z])/g,"$1 $2").replace(/_/g," ").replace(/-/g," ").split(" ").map(y=>y.charAt(0).toUpperCase()+y.slice(1)).join(" "),"prettyString");var k=o(e=>{let{resource:t,field:c,i18n:a}=e;if(a){let p=`translate("${t.name}.fields.${c.key}")`;return e.noBraces?`${p}`:`{${p}}`}let s=U(c.key);return e.noQuotes?s:`"${s}"`},"translatePrettyString");var Qe=K(require("pluralize"));var W=o((e="")=>Qe.default.isSingular(e)?e:Qe.default.singular(e),"toSingular");var Ke=K(require("pluralize"));var re=o((e="")=>Ke.default.isPlural(e)?e:Ke.default.plural(e),"toPlural");var ee=o(e=>{var p;let{resource:t,action:c,i18n:a,noBraces:s}=e;if(a){let y=`translate("${t.name}.titles.${c}")`;return s?y:`{${y}}`}return c==="list"?U(re(t.label??((p=t==null?void 0:t.meta)==null?void 0:p.label)??t.name)):`${U(W(t.name))} ${U(c)}`},"translateActionTitle");var ne=o(e=>{let{action:t,i18n:c,noQuotes:a}=e;return c?`{translate("buttons.${t}")}`:a?U(t):`"${U(t)}"`},"translateButtonTitle");var _e=o(e=>e.replace(/export\s+(const|let|var|type|interface|function|class)\s+(\w+)\s*(=|:)\s*/g,"$1 $2 = ").split(`
`).filter(s=>!s.trim().startsWith("export default")).join(`
`),"replaceExports");var yt=/import(?:(?:(?:[ \n\t]+([^ *\n\t\{\},]+)[ \n\t]*(?:,|[ \n\t]+))?([ \n\t]*\{(?:[ \n\t]*[^ \n\t"'\{\}]+[ \n\t]*,?)+\})?[ \n\t]*)|[ \n\t]*\*[ \n\t]*as[ \n\t]+([^ \n\t\{\}]+)[ \n\t]+)from[ \n\t]*(?:['"])([^'"\n]+)(?:['"])(?:;?)/g,ir=/import[ \n\t](?:['"])([^'"\n]+)(?:['"])(?:;?)/g,cr=/((?:\w|\s|_)*)( as )((?:\w|\s|_)*)( |,)?/g,We=o((e,t={})=>{let c=e.matchAll(yt),a=new Set;for(let s of c){let[,p,y,u,f]=s;if(f in t){let I=t[f];if(f==="react")continue;p&&a.add(`const { default: ${p} } = ${I};`),y&&a.add(`const${y.replace(cr,"$1: $3$4")} = ${I};`),u&&a.add(`const ${u} = ${I};`)}}return ft(`
    ${Array.from(a).join(`
`)}
    ${e.replace(yt,"").replace(ir,"")}
    `)},"replaceImports");var Ce=K(require("pluralize"));var qe=o((e,t)=>e?t.find(a=>{let s=z(e.key);return a.name===s||a.name===(Ce.default.isPlural(s)?s:Ce.default.plural(s))||a.name===(Ce.default.isSingular(s)?s:Ce.default.singular(s))}):void 0,"resourceFromInferred");var ht=K(require("pluralize"));var oe=o((e,t,c="Inferred")=>{let a=e.replace(/[^a-zA-Z0-9]/g," "),s=ht.default.singular(a),y=U(s).replace(/ /g,"")||c,u=t.charAt(0).toUpperCase()+t.slice(1);return/^[a-zA-Z]/.test(y)?`${y}${u}`:`${u}${y}`},"componentName");var pr=o(e=>e?Array.isArray(e)?e.join(""):typeof e=="string"?e:typeof e=="object"||typeof e=="function"?e.toString():typeof e=="number"?String(e):typeof e=="boolean"&&e?"true":e:"","handleExpression"),x=o((e,...t)=>e.reduce((a,s,p)=>`${a}${s}${pr(t[p])}`,"").trim(),"jsx");var Ge=o((e,t)=>`
    ${e??""}
    
    render(typeof ${t} !== "undefined" ? <${t} /> : <></>);
    `,"prepareLiveCode");var Ze=o(e=>e==null?void 0:e.replace(/(\/\* hidden-start \*\/)(.|\n)*?(\/\* hidden-end \*\/)/g,""),"removeHiddenCode");var ur=/^[a-zA-Z_$][a-zA-Z_$0-9]*$/,Ue=o(e=>ur.test(e),"shouldDotAccess");function Ye(e,t,c){let a=`${e}`;return t&&(a+="?.",Ue(t)?a+=t:a+=`['${t}']`),c&&(a+="?.",Ue(c)?a+=c:a+=`['${c}']`),a}o(Ye,"accessorSingle");function lr(e,t,c){return(c??[]).map(s=>Ye(e,t,s))}o(lr,"accessorMultiple");var d=o((e,t,c,a=' + " " + ')=>Array.isArray(c)?a?lr(e,t,c).join(a):Ye(e,t,c[0]):Ye(e,t,c),"accessor"),q=o((e,t,c)=>{let a=e;return t&&(a+=`.${t}`),c&&(Array.isArray(c)?a+=`.${c[0]}`:a+=`.${c}`),a},"dotAccessor");var ie=o(e=>{let t=e.reduce((a,[s,p,y])=>(a[p]||(a[p]=[]),a[p].includes(s)||(y?a[p].push([s,!0]):a[p].push(s)),a),{});return Object.entries(t).map(([a,s])=>{let p=s.find(I=>Array.isArray(I)&&I[1]),y=s.filter(I=>typeof I=="string"),u=p?`${p[0]}`:"",f=y.length?`{ ${y.join(", ")} }`:"";return`import ${u}${u&&f?", ":""}${f} from "${a}";`})},"printImports");var mr=/^id$/i,Fe=o(e=>mr.test(e),"isIDKey");var we=o(e=>e.relationInfer&&e.relationInfer.accessor?Array.isArray(e.relationInfer.accessor)?`optionLabel: "${e.relationInfer.accessor[0]}",`:e.relationInfer.accessor!=="title"?`optionLabel: "${e.relationInfer.accessor}",`:"":"","getOptionLabel");var ce=o(e=>{},"noOp");var P=o((e,t="")=>{if(Number.isInteger(Number(e)))return t.charAt(0).toLowerCase()+t.slice(1)+e;let a=e.replace(/^\d+/,"").replace(/[^a-zA-Z0-9]/g,"-"),s=U(a).replace(/ /g,"");return`${s.charAt(0).toLowerCase()+s.slice(1)}${t.charAt(0).toUpperCase()+t.slice(1)}`},"getVariableName");var D=o((e,t,c)=>{if(t&&c&&e){let a=e?t[e]:{},s=(a==null?void 0:a[c])??(a==null?void 0:a.default);return s?`meta: ${JSON.stringify(s)},`:""}return""},"getMetaProps"),se=o((e,t,c)=>{if(t&&c&&e){let a=e?t[e]:{},p=[...c,"default"].reduce((y,u)=>y??(a==null?void 0:a[u]),void 0);if(p)return p}},"pickMeta");var Se=o(e=>e===void 0?e:typeof e=="string"?`"${e}"`:e,"idQuoteWrapper");var $t=o(e=>{let t=o((c,a,s,p=t,y)=>{let u=e.map(I=>I(c,a,s,p,y));return je(u)},"fieldInferencer");return t},"composeInferencers");var xt=o(e=>o((c,a,s,p,y,u)=>e.reduce((f,I)=>I(f,a,s,p,y,u),c),"fieldTransformer"),"composeTransformers");var It=o((e,t,c,a,s)=>{let p=Array.isArray(t),y=Array.isArray(t)&&t.every(u=>typeof u=="string"||typeof u=="number");if(p){if(!y){let f=a(e,t[0],c,a,s);return f?{...f,key:e,multiple:!0,priority:1}:!1}let u=a(e,t[0],c,a,s)||{type:"string"};return{key:e,multiple:!0,priority:1,type:u==null?void 0:u.type}}return!1},"arrayInfer");var vt=o((e,t)=>typeof t=="boolean"?{key:e,type:"boolean"}:!1,"booleanInfer");var Je=K(require("dayjs"));var dr=/(_at|_on|At|On|AT|ON)(\[\])?$/,fr=["/",":","-","."],bt=o((e,t)=>{let c=dr.test(e)&&(0,Je.default)(t).isValid(),a=typeof t=="string"&&(0,Je.default)(t).isValid();return typeof t=="string"&&fr.some(p=>t.includes(p))&&(c||a)?{key:e,type:"date",priority:1}:!1},"dateInfer");var gr=/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,kt=o((e,t)=>typeof t=="string"&&gr.test(t)?{key:e,type:"email",priority:1}:!1,"emailInfer");var yr=/\.(gif|jpe?g|tiff?|png|webp|bmp|svg)$/i,Ct=o((e,t)=>typeof t=="string"&&yr.test(t)?{key:e,type:"image",priority:2}:!1,"imageInfer");var Ft=o((e,t)=>typeof t>"u"||t===null?null:!1,"nullishInfer");var wt=o((e,t)=>{let c=typeof t=="string"&&t.length>0,a=!isNaN(t);return c&&a||typeof t=="number"?{key:e,type:"number"}:!1},"numberInfer");var St=/id$/i,At=o((e,t,c,a,s)=>{let p=t!==null,y=!Array.isArray(t);if(p&&y&&typeof t=="object"){let f=Object.keys(t).length===1&&St.test(Object.keys(t)[0]),I=Object.keys(t).some(C=>St.test(C));if(f)return{key:e,type:"relation",relation:!0,accessor:"id",priority:1};let v=De(e,t);if((v?v.length>0:!1)&&v){let C=v&&Array.isArray(v)?v[0]:v,$=a(C,t[C],t,a,s);if($){let A=Array.isArray(v)?v.map(l=>$.accessor?`${l}.${Array.isArray($.accessor)?$.accessor[0]:$.accessor}`:l):$.accessor?Array.isArray($.accessor)?`${v}.${$.accessor[0]}`:`${v}.${$.accessor}`:v;return($==null?void 0:$.type)==="text"&&(s==="create"||s==="edit")&&I?{key:e,type:"relation",relation:!0,accessor:"id",priority:1}:{...$,fieldable:!0,key:e,accessor:A,priority:1}}}return{key:e,fieldable:!1,type:"object",priority:1}}return!1},"objectInfer");var Rt=o((e,t)=>typeof t=="string"&&t.length>100?{key:e,type:"richtext",priority:1}:!1,"richtextInfer");var Pt=o((e,t)=>typeof t=="string"?{key:e,type:"text"}:!1,"textInfer");var hr=/^(https?|ftp):\/\/(-\.)?([^\s/?\.#-]+\.?)+(\/[^\s]*)?$/i,Tt=o((e,t)=>typeof t=="string"&&hr.test(t)?{key:e,type:"url",priority:1}:!1,"urlInfer");var Nt=[It,vt,bt,kt,Ct,Ft,wt,At,gt,Rt,Pt,Tt];var Lt=o((e,t,c,a)=>e.map(p=>{if(!p.relation&&(p.type==="text"||p.type==="richtext"||p.type==="number")&&!p.canRelation){let y=/^[a-z0-9-]+$/,u=o(C=>y.test(`${C}`),"isValidUUID"),f=p.key.toLowerCase()!=="id";return!(!p.accessor||typeof p.accessor=="string")||!f?p:(p.multiple?a[p.key]:[a[p.key]]).every(C=>u(p.accessor?C[p.accessor]:C))?{...p,canRelation:!0}:p}return p}),"basicToRelation");var $r=/(image|photo|avatar|cover|thumbnail|icon)/i,Vt=o(e=>e.map(c=>c.type==="url"&&$r.test(c.key)?{...c,type:"image"}:c),"imageByKey");var Et=o((e,t)=>e.map(a=>{let s=qe(a,t);return s?{...a,relation:!0,type:"relation",resource:s}:a}),"relationByResource");var Mt=o((e,t,c,a,s,p)=>e.map(u=>{if(u.relation&&u.type==="relation"&&!u.resource){let f=u.accessor?a[u.key][u.accessor]:a[u.key],I=s(u.key,f,a,s,p);if(I&&I.type!=="relation")return{...u,fieldable:!0,relation:!1,type:I.type,canRelation:!0}}return u}),"relationToFieldable");var Ot=[Vt,Et,Mt,Lt];var te=K(require("react")),xr=K(require("@refinedev/core")),Pe=require("@aliemir/react-live");var Dt=[["react","React",te.default],["@refinedev/core","RefineCore",xr]],Xe=te.default.createContext({}),et=o(({fetchError:e,code:t,additionalScope:c,errorComponent:a})=>{let s=(t&&_e(We(t,[...Dt,...c??[]].reduce((u,[f,I,v,T])=>T?u:{...u,[f]:I},{}))))??"",p=te.default.useMemo(()=>({React:te.default,...te.default,...[...Dt,...c??[]].reduce((u,[f,I,v])=>({...u,[I]:v}),{})}),[c]),y=te.default.useMemo(()=>o(()=>{let{error:f}=te.default.useContext(Xe);return a?te.default.createElement(a,{error:f?`<p>An error occured while rendering the generated component.You can check the generated code from the below "Show Code" button and fix the error manually.</p>
                                    <p>If you think this is a bug, please report the issue at <a target="_blank" rel="noopener noreferrer" href="https://github.com/refinedev/refine/issues">https://github.com/refinedev/refine/issues</a></p>
                                    <p>Exception:</p>
                                    <code>${f}</code>`:typeof e=="string"?e:e?`<p>Something went wrong while fetching the resource data.</p>
                                    <p>To learn more about the Inferencer, please check the <a href="https://refine.dev/docs/packages/documentation/inferencer/" target="_blank">documentation</a>.</p>`:void 0}):null},"LiveErrorComponent"),[a,e]);return te.default.createElement(Pe.LiveProvider,{Context:Xe,code:s,scope:p,noInline:!0},!e&&te.default.createElement(Pe.LivePreview,{Context:Xe}),te.default.createElement(y,null))},"LiveComponent");var g=K(require("react")),Te=K(require("prism-react-renderer")),jt=K(require("prism-react-renderer/themes/vsDark"));var le=o(({code:e,loading:t})=>{let c=g.default.useMemo(()=>Be(e??""),[e]),[a,s]=g.default.useState(!1),[p,y]=g.default.useState(!1),[u,f]=g.default.useState(!1),[I,v]=g.default.useState(!1),[T,C]=g.default.useState(!1);return g.default.useEffect(()=>{if(!t){let $=setTimeout(()=>{s(!0)},300);return()=>{clearTimeout($)}}return()=>{}},[t]),g.default.useEffect(()=>{if(typeof window<"u"){let $=window.matchMedia("(max-width: 449px)");$.matches?f(!1):f(!0);let A=o(()=>{$.matches?f(!1):f(!0)},"handleResize");return window.addEventListener("resize",A),()=>{window.removeEventListener("resize",A)}}return()=>{}},[]),g.default.useEffect(()=>{if(typeof window<"u"){let $=window.matchMedia("(max-width: 1280px)");$.matches?v(!0):v(!1);let A=o(()=>{$.matches?v(!0):v(!1)},"handleResize");return window.addEventListener("resize",A),()=>{window.removeEventListener("resize",A)}}return()=>{}},[]),c&&!t?g.default.createElement(g.default.Fragment,null,u&&g.default.createElement("div",{className:"refine-inferencer--code-viewer",style:{position:"sticky",bottom:"24px",paddingTop:"24px",left:0,right:0,width:"100%",zIndex:10,display:"flex",justifyContent:"center",transition:"all 0.2s ease",opacity:a&&!p?1:0,transform:a&&!p?"translateY(0)":"translateY(100px)"}},g.default.createElement("div",{style:{width:"calc(100% - calc(64px * 2))",maxWidth:"1080px",padding:"20px",backgroundColor:"#1A1A1A",boxShadow:"0px 4px 16px -4px rgba(0, 0, 0, 0.5), 0px 8px 32px -8px rgba(0, 0, 0, 0.35)",borderRadius:"16px",display:"flex",alignItems:"center",justifyContent:"space-between",flexDirection:I?"column":"row",gap:"12px"}},g.default.createElement("div",{style:{display:"flex",alignItems:"flex-start",gap:"8px"}},g.default.createElement("div",null,g.default.createElement(vr,null)),g.default.createElement("div",{style:{fontSize:"14px",lineHeight:"20px",color:"#ffffff"}},g.default.createElement("p",{style:{padding:0,margin:0}},"Most of the page code is auto-generated by the"," ",g.default.createElement("span",{style:{textDecoration:"underline"}},"Inferencer")," ","feature, based on your backend data structure."),g.default.createElement("p",{style:{padding:0,margin:0}},"While this is an excellent way to experiment with refine,"," ",g.default.createElement("span",{style:{fontWeight:600}},"it's not intended to be used on production.")))),g.default.createElement("div",null,g.default.createElement("button",{onPointerEnter:()=>C(!0),onPointerLeave:()=>C(!1),onClick:()=>y(!0),style:{appearance:"none",border:"none",padding:"10px 16px",borderRadius:"4px",display:"flex",alignItems:"center",justifyContent:"center",background:"#0080FF",color:"#ffffff",fontSize:"14px",lineHeight:"20px",fontWeight:600,gap:"8px",cursor:"pointer",transition:"all 0.2s ease",transform:T?"scale(1.025)":void 0,filter:T?"brightness(1.1)":void 0}},g.default.createElement(Bt,{style:{flexShrink:0}}),g.default.createElement("span",null,"Show the auto-generated code"))))),g.default.createElement(Ir,{visible:p,onClose:()=>y(!1),code:c})):null},"SharedCodeViewer"),Ir=o(({visible:e,onClose:t,code:c=""})=>{let a=g.default.useRef(null),[s,p]=g.default.useState(!1),[y,u]=g.default.useState(!1),[f,I]=g.default.useState(!1),[v,T]=g.default.useState(!1);g.default.useEffect(()=>{if(typeof document<"u"){let m=o(i=>{a.current&&!a.current.contains(i.target)&&t()},"onOutsideClick");return document.addEventListener("pointerdown",m),()=>{document.removeEventListener("pointerdown",m)}}return()=>{}},[t]);let C=o(()=>{typeof navigator<"u"&&(navigator.clipboard.writeText(c),p(!0),setTimeout(()=>{p(!1)},1e3))},"onCopy"),$=g.default.createElement("div",{style:{display:"flex",justifyContent:"space-between",alignItems:"center"}},g.default.createElement("div",{style:{fontWeight:700,fontSize:"20px",lineHeight:"32px",color:"#0D0D0D"}},"Auto-generated code by Inferencer"),g.default.createElement("button",{onClick:t,onPointerEnter:()=>I(!0),onPointerLeave:()=>I(!1),style:{flexShrink:0,appearance:"none",border:"none",background:"none",padding:0,margin:0,outline:"none",borderRadius:"50px",width:"32px",height:"32px",cursor:"pointer",transition:"all 0.15s ease",transform:f?"scale(1.05)":void 0,filter:f?"brightness(0.8)":void 0}},g.default.createElement(br,null))),A=g.default.createElement("div",{style:{fontSize:"13px",borderRadius:"8px",flex:"1",overflow:"scroll",background:"#1E1E1E"}},g.default.createElement(Te.default,{...Te.defaultProps,theme:jt.default,code:c,language:"tsx"},({className:m,style:i,tokens:b,getLineProps:N,getTokenProps:S})=>g.default.createElement("pre",{className:m,style:{...i,padding:"14px 14px 14px 14px",margin:"0",width:"100%",boxSizing:"border-box"}},b.map((B,_)=>g.default.createElement("div",{key:_,...N({line:B,key:_})},B.map((R,M)=>g.default.createElement("span",{key:M,...S({token:R,key:M})})))))));return g.default.createElement("div",{className:"refine-inferencer--code-viewer-modal",style:{position:"fixed",top:0,left:0,right:0,bottom:0,zIndex:9999,backgroundColor:"rgba(0, 0, 0, 0.5)",transition:"all 0.2s ease",opacity:e?1:0,pointerEvents:e?"all":"none",display:"flex",justifyContent:"center",alignItems:"center"}},g.default.createElement("div",{ref:a,style:{transform:e?"scale(1) translateY(0px)":"scale(0) translateY(-200px)",transition:"all 0.25s cubic-bezier(.35,1.29,.81,1.08)",transitionDelay:"0.25",width:"calc(100% - calc(32px * 2))",height:"calc(100% - calc(32px * 2))",backgroundColor:"#fff",maxWidth:"640px",maxHeight:"720px",borderRadius:"8px",padding:"16px"}},g.default.createElement("div",{style:{display:"flex",flexDirection:"column",height:"100%",width:"100%",gap:"16px"}},$,A,g.default.createElement("div",{style:{display:"flex",justifyContent:"flex-start",alignItems:"flex-end",gap:"16px"}},g.default.createElement("button",{onPointerEnter:()=>u(!0),onPointerLeave:()=>u(!1),onClick:C,style:{appearance:"none",height:"40px",border:"none",padding:"10px 16px",borderRadius:"4px",display:"flex",alignItems:"center",justifyContent:"center",background:"#0080FF",color:"#ffffff",fontSize:"14px",lineHeight:"20px",fontWeight:600,gap:"8px",cursor:"pointer",position:"relative",overflow:"hidden",transition:"filter 0.2s ease",transform:y?"scale(1.025)":"scale(1)",filter:y?"brightness(1.1)":void 0}},g.default.createElement(Ht,{style:{flexShrink:0,marginTop:"-2px",marginBottom:"-2px"}}),g.default.createElement("span",null,"Copy Generated Code"),g.default.createElement("div",{style:{position:"absolute",width:"100%",height:"100%",top:0,left:0,display:"flex",alignItems:"center",justifyContent:"center",background:"#0080FF",transition:"all 0.2s ease",transform:s?"translateY(0)":"translateY(40px)"}},g.default.createElement(Ht,{style:{flexShrink:0,marginTop:"-2px",marginBottom:"-2px"}}),g.default.createElement("span",null,"Copied!"))),g.default.createElement("a",{onPointerEnter:()=>T(!0),onPointerLeave:()=>T(!1),href:"https://refine.dev/docs/packages/documentation/inferencer",target:"_blank",rel:"noreferrer",style:{appearance:"none",textDecoration:"none",border:"none",padding:"10px 16px",borderRadius:"4px",display:"flex",alignItems:"center",justifyContent:"center",background:"rgba(0, 128, 255, 0.1)",color:"#0080FF",fontSize:"14px",lineHeight:"20px",fontWeight:600,gap:"8px",cursor:"pointer",transition:"all 0.2s ease",transform:v?"scale(1.025)":void 0,filter:v?"brightness(1.1)":void 0}},g.default.createElement(Bt,{style:{flexShrink:0}}),g.default.createElement("span",null,"Learn more about inferencer"))))))},"CodeModal"),vr=o(e=>g.default.createElement("svg",{xmlns:"http://www.w3.org/2000/svg",width:20,height:20,fill:"none",...e},g.default.createElement("path",{fill:"#0080FF",fillRule:"evenodd",d:"M10 20C4.477 20 0 15.523 0 10S4.477 0 10 0s10 4.477 10 10-4.477 10-10 10Zm0-15a1.25 1.25 0 1 0 0 2.5A1.25 1.25 0 0 0 10 5Zm0 10c.69 0 1.25-.56 1.25-1.25V10a1.25 1.25 0 1 0-2.5 0v3.75c0 .69.56 1.25 1.25 1.25Z",clipRule:"evenodd"})),"InfoIcon"),Bt=o(e=>g.default.createElement("svg",{xmlns:"http://www.w3.org/2000/svg",width:16,height:16,fill:"none",...e},g.default.createElement("path",{fill:"currentColor",d:"M5 2a1 1 0 0 1 0 2H4v8h8v-1a1 1 0 1 1 2 0v1a2 2 0 0 1-2 2H4a2 2 0 0 1-2-2V4a2 2 0 0 1 2-2h1Z"}),g.default.createElement("path",{fill:"currentColor",d:"M9 2a1 1 0 0 0 0 2h1.586L6.293 8.293a1 1 0 0 0 1.414 1.414L12 5.414V7a1 1 0 1 0 2 0V3a1 1 0 0 0-1-1H9Z"})),"OpenIcon"),br=o(e=>g.default.createElement("svg",{xmlns:"http://www.w3.org/2000/svg",width:32,height:32,fill:"none",...e},g.default.createElement("path",{fill:"#A6A6A6",fillRule:"evenodd",d:"M16 32C7.163 32 0 24.837 0 16S7.163 0 16 0s16 7.163 16 16-7.163 16-16 16Zm-2.586-21.414a2 2 0 1 0-2.828 2.828L13.172 16l-2.586 2.586a2 2 0 1 0 2.828 2.828L16 18.828l2.586 2.586a2 2 0 1 0 2.828-2.828L18.828 16l2.586-2.586a2 2 0 1 0-2.828-2.828L16 13.172l-2.586-2.586Z",clipRule:"evenodd"})),"CloseIcon"),Ht=o(e=>g.default.createElement("svg",{xmlns:"http://www.w3.org/2000/svg",width:20,height:20,viewBox:"0 0 24 24",fill:"none",...e},g.default.createElement("path",{fill:"currentColor",fillRule:"evenodd",d:"M8 5a3 3 0 0 0 3 3h2a3 3 0 0 0 3-3h1a2 2 0 0 1 2 2v11a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V7a2 2 0 0 1 2-2h1Zm0 7a1 1 0 0 1 1-1h6a1 1 0 1 1 0 2H9a1 1 0 0 1-1-1Zm1 3a1 1 0 1 0 0 2h4a1 1 0 1 0 0-2H9Z",clipRule:"evenodd"}),g.default.createElement("path",{fill:"currentColor",d:"M13 4a1 1 0 1 1 0 2h-2a1 1 0 1 1 0-2h2Z"})),"ClipboardIcon");var me=K(require("react")),Ne=require("@refinedev/core");var zt=o((e,t,c,a)=>{let{resource:s,id:p,resources:y}=(0,Ne.useResource)(t),u=c??p,f=(0,Ne.useDataProvider)(),[I,v]=me.default.useState(void 0),[T,C]=me.default.useState(void 0),[$,A]=me.default.useState(void 0),[l,m]=me.default.useState(!0),[i,b]=me.default.useState(!1),N=me.default.useCallback(async S=>{var R;let B=ke(s)??He(s==null?void 0:s.name,void 0,y),_=f(B);b(!0),v(void 0);try{if((e==="list"||e==="create")&&s){let M=await _.getList({resource:s==null?void 0:s.name,meta:se((s==null?void 0:s.identifier)??(s==null?void 0:s.name),a,["getList"])}),n=(R=M.data)==null?void 0:R[0];n||v(`<p>No records/data found for resource "${s==null?void 0:s.name}".</p>
                            <p>Please check your data provider and resource.</p>
                            <p>For more info, please check the <a href="https://refine.dev/docs/packages/documentation/inferencer/" target="_blank">documentation</a>.</p>`),C(n),A(M.data),setTimeout(()=>{b(!1)},500)}if((e==="edit"||e==="show")&&S&&s){let n=(await _.getOne({resource:s==null?void 0:s.name,id:S,meta:se((s==null?void 0:s.identifier)??(s==null?void 0:s.name),a,["getOne"])})).data;n||v(`<p>No records/data found for resource "${s==null?void 0:s.name}".</p>
                            <p>Please check your data provider and resource.</p>
                            <p>For more info, please check the <a href="https://refine.dev/docs/packages/documentation/inferencer/" target="_blank">documentation</a>.</p>`),C(n),setTimeout(()=>{b(!1)},500)}}catch(M){console.warn("An error occured while fetching the resource data. Please check the error message below:",M),v(`<p>Something went wrong while fetching the resource data.</p>
                    <p>Please check your data provider and API for resource "${s==null?void 0:s.name}".</p>
                    <p>For more info, please check the <a href="https://refine.dev/docs/packages/documentation/inferencer/" target="_blank">documentation</a>.</p>`),setTimeout(()=>{b(!1)},500)}},[e,f,s,y]);return me.default.useEffect(()=>{m(!1),!i&&!T&&N(u)},[N,u]),{data:T,datas:$,loading:i,initial:l,error:I}},"useInferFetch");var he=K(require("react")),Qt=require("@refinedev/core");var tt=require("lodash");var Kt=o(({record:e,fields:t,infer:c,meta:a})=>{let s=(0,Qt.useDataProvider)(),[p,y]=he.default.useState([]),[u,f]=he.default.useState(!0),[I,v]=he.default.useState(!1),T=he.default.useCallback(async C=>{console.groupCollapsed("@refinedev/inferencer is trying to detect relations");let $=[];v(!0);try{let A=C.map(async m=>{var i,b,N,S,B;if(m&&(m.relation||m.canRelation)&&e){if(m.relationInfer)return m;let _=ke(m.resource),R=s(_),M=m.multiple,n=Array.isArray(m.accessor)?void 0:m.multiple?e[m.key].map(r=>m.accessor?(0,tt.get)(r,m.accessor):r)[0]:m.accessor?(0,tt.get)(e[m.key],m.accessor):e[m.key];if(n&&m.resource)try{let r={};if(M&&R.getMany){let{data:F}=await R.getMany({resource:m.resource.name,ids:[n],meta:se(((i=m.resource)==null?void 0:i.identifier)??((b=m.resource)==null?void 0:b.name),a,["getMany"])});r=F==null?void 0:F[0]}else{let{data:F}=await R.getOne({resource:m.resource.name,id:n,meta:se(((N=m.resource)==null?void 0:N.identifier)??((S=m.resource)==null?void 0:S.name),a,M?["getMany","getOne"]:["getOne"])});r=F}$.push({status:"success",resource:m.resource.name,field:m.key});let h=c("__",r,{},c);return{...m,relationInfer:h}}catch{return $.push({status:"error",resource:m.resource.name,field:m.key}),{...m,relationInfer:null}}if(n){let r,h;try{let H={};if(M&&R.getMany){let{data:L}=await((B=R.getMany)==null?void 0:B.call(R,{resource:re(z(m.key)),ids:[n],meta:se(re(z(m.key)),a,["getMany"])}));H=L==null?void 0:L[0]}else{let{data:L}=await R.getOne({resource:re(z(m.key)),id:n,meta:se(re(z(m.key)),a,M?["getMany","getOne"]:["getOne"])});H=L}$.push({status:"success",resource:re(z(m.key)),field:m.key}),r=H,h=!0}catch{$.push({status:"error",resource:re(z(m.key)),field:m.key});let L={};try{if(M&&R.getMany){let{data:Q}=await R.getMany({resource:W(z(m.key)),meta:se(W(z(m.key)),a,["getMany"]),ids:[n]});L=Q==null?void 0:Q[0]}else{let{data:Q}=await R.getOne({resource:W(z(m.key)),meta:se(W(z(m.key)),a,M?["getMany","getOne"]:["getOne"]),id:n});L=Q}$.push({status:"success",resource:W(z(m.key)),field:m.key}),r=L,h=!1}catch{return $.push({status:"error",resource:W(z(m.key)),field:m.key}),{...m,relationInfer:null}}}let F=c("__",r,{},c),w=z(m.key);return{...m,relation:!0,type:"relation",resource:{name:h?re(w):W(w)},fieldable:!1,canRelation:void 0,relationInfer:F}}return{...m,relationInfer:null}}return m}),l=await Promise.all(A);y(l.filter(m=>m)),setTimeout(()=>{v(!1)},500)}catch{setTimeout(()=>{v(!1)},500)}setTimeout(()=>{console.log(`Tried to detect relations with ${$.length} attempts and succeeded with ${$.filter(A=>A.status==="success").length} attempts.`),console.groupEnd(),console.info("@refinedev/inferencer may send multiple requests to your API for nonexistent resources when trying to detect relations. To learn more about how the inferencer works, visit https://s.refine.dev/how-inferencer-works")},500)},[s,e]);return he.default.useEffect(()=>{f(!1),!I&&t&&t.length>0&&p.length===0&&T(t)},[T,I,t,T]),{fields:p,loading:I,initial:u}},"useRelationFetch");var de=o(({type:e,additionalScope:t=[],customElements:c=[],fieldTransformers:a=[],renderer:s,loadingComponent:p,errorComponent:y,codeViewerComponent:u})=>{let f=$t([...Nt,...c]),I=xt([...Ot,...a]),v=o(({resourceName:C,fieldTransformer:$,hideCodeViewerInProduction:A,meta:l,id:m})=>{var ct,pt;let{resource:i,resources:b}=(0,$e.useResource)(C),{i18nProvider:N}=(0,G.useContext)($e.TranslationContext),{resource:S}=(0,$e.useResource)(),{data:B,datas:_,loading:R,initial:M,error:n}=zt(e,C??(i==null?void 0:i.name),m,l),r=o((V,E,Y)=>{let J=f(V,E,Y,f,e);if(J&&i){let X=I([J],b,i,Y,f,e),j=$?X.flatMap(O=>{let Z=$(O);return Z?[Z]:[]}):X;return j==null?void 0:j[0]}},"inferSingleField"),h=o(V=>Object.keys(V).map(Y=>{let J=V[Y];return r(Y,J,V)}).filter(Boolean),"inferSingleRecord"),F=o(V=>{let E=V.map(O=>h(O)),Y=V.flatMap(O=>Object.keys(O)).filter((O,Z,Ie)=>Ie.indexOf(O)===Z),J={},X=Y.map(O=>{let Z=E.map(pe=>pe.find(ae=>ae.key===O));return Z.reduce((pe,ae,Re)=>{if(!ae)return pe;let ve=Z.filter(ue=>(ue==null?void 0:ue.key)===ae.key&&(ue==null?void 0:ue.type)===ae.type).length;return ve>pe.count?(J[O]=V[Re][O],{count:ve,field:ae}):pe},{count:0,field:void 0}).field});return{commonRecord:J,inferredFields:X}},"inferMultipleRecords"),[w,H]=G.default.useMemo(()=>{if(_&&(e==="list"||e==="create")){let V=F(_);return[V.inferredFields,V.commonRecord]}if(B){let V=Object.keys(B).map(E=>{let Y=B[E];return f(E,Y,B,f,e)}).filter(Boolean);if(i){let E=I(V,b,i,B,f,e);return[$?E.flatMap(J=>{let X=$(J);return X?[X]:[]}):E,B]}return[[],B]}return[[],void 0]},[B,_,b,i,$]),{fields:L,loading:Q}=Kt({record:H,fields:w,infer:f,meta:l}),at=G.default.useMemo(()=>{let V=[];return L.forEach((E,Y,J)=>{if(E.resource){if(V.findIndex(j=>{var O,Z;return((O=j.resource)==null?void 0:O.name)===((Z=E.resource)==null?void 0:Z.name)})>-1)return;let X=J.filter((j,O)=>{var Z,Ie,pe,ae,Re,ve;if(O!==Y){let ue=E.resource,Yt=j.resource,Jt=((Z=j.resource)==null?void 0:Z.identifier)||((Ie=E.resource)==null?void 0:Ie.identifier)?((pe=j.resource)==null?void 0:pe.identifier)===((ae=E.resource)==null?void 0:ae.identifier):((Re=j.resource)==null?void 0:Re.name)===((ve=E.resource)==null?void 0:ve.name);return ue&&Yt&&Jt}else return!1});if(X.length>0)if(e==="create"||e==="edit"){let j;[E,...X].find(O=>{O.fieldable!==!0&&j===void 0&&(j=O)}),j?V.push(j):V.push(E)}else{let j;[E,...X].find(O=>{O.fieldable!==!1&&j===void 0&&(j=O)}),j?V.push(j):V.push(E)}else V.push(E)}else V.push(E)}),V},[L,e]),it=G.default.useMemo(()=>!R&&!Q&&!M&&i?s({resource:i,resources:b,fields:at,infer:f,meta:l,isCustomPage:i.name!==(S==null?void 0:S.name),id:m,i18n:!!N}):"",[i,b,at,R,Q]),Ut=A;return G.default.createElement(G.default.Fragment,null,p&&(R||Q)&&G.default.createElement(p,null),!R&&!Q&&G.default.createElement(G.default.Fragment,null,G.default.createElement(et,{fetchError:!R&&n?n:!R&&!M&&!B,code:Ge(it,oe(((ct=i==null?void 0:i.meta)==null?void 0:ct.label)??((pt=i==null?void 0:i.options)==null?void 0:pt.label)??(i==null?void 0:i.label)??(i==null?void 0:i.name)??"Resource",e)),errorComponent:y,additionalScope:t}),typeof u<"u"&&!Ut?G.default.createElement(u,{code:Ze(it),loading:R||Q}):null))},"Inferencer");return o(({name:C,resource:$,fieldTransformer:A,meta:l,hideCodeViewerInProduction:m,id:i})=>{let{resource:b}=(0,$e.useResource)($??C),N=`${(b==null?void 0:b.identifier)??(b==null?void 0:b.name)}-${e}-${i}`;return G.default.createElement(v,{hideCodeViewerInProduction:m,fieldTransformer:A,resourceName:$??C,meta:l??{},key:N,id:i})},"InferencerComponent")},"createInferencer");var Le=K(require("react"));var fe=o(({error:e})=>e?Le.default.createElement("div",{style:{display:"flex",justifyContent:"center",alignItems:"center",minHeight:"300px",position:"relative"}},Le.default.createElement("div",{style:{position:"absolute",top:"50%",left:"50%",transform:"translate(-50%, -50%)"}},Le.default.createElement("div",{style:{fontSize:"1.5rem",fontWeight:"bold"},dangerouslySetInnerHTML:{__html:e??""}}))):null,"ErrorComponent");var xe=K(require("react"));var ge=o(()=>{let[e,t]=xe.default.useState(0);return xe.default.useEffect(()=>{if(typeof window<"u"){let c=setInterval(()=>{t(a=>a===3?0:a+1)},300);return()=>{clearInterval(c)}}return()=>{}},[]),xe.default.createElement("div",{style:{display:"flex",justifyContent:"center",alignItems:"center",minHeight:"120px",position:"relative"}},xe.default.createElement("div",{style:{position:"absolute",top:"50%",left:"50%",transform:"translate(-50%, -50%)"}},xe.default.createElement("span",{style:{fontSize:"1rem",fontWeight:600}},"loading",".".repeat(e))))},"LoadingComponent");var rt=o(({resource:e,fields:t,meta:c,isCustomPage:a,id:s,i18n:p})=>{let y=oe(e.label??e.name,"show"),u="record",f=[["React","react",!0],["useShow","@refinedev/core"],["useResource","@refinedev/core"],["useNavigation","@refinedev/core"],["IResourceComponentsProps","@refinedev/core"]];p&&f.push(["useTranslate","@refinedev/core"]);let v=t.filter(r=>(r==null?void 0:r.relation)&&!(r!=null&&r.fieldable)&&(r==null?void 0:r.resource)).filter(Boolean).map(r=>{var h,F,w,H;if(r!=null&&r.relation&&!r.fieldable&&r.resource){if(r.multiple){f.push(["useMany","@refinedev/core"]);let L=d(u,r.key);return r.accessor&&(L=`${d(u,r.key)}?.map((item: any) => ${d("item",void 0,r.accessor)})`),`
                const { data: ${P(r.key,"Data")}, isLoading: ${P(r.key,"IsLoading")} } =
                useMany({
                    resource: "${r.resource.name}",
                    ids: ${L} || [],
                    queryOptions: {
                        enabled: !!${u} && !!${L}?.length,
                    },
                    ${D(((h=r==null?void 0:r.resource)==null?void 0:h.identifier)??((F=r==null?void 0:r.resource)==null?void 0:F.name),c,"getMany")}
                });
                `}return f.push(["useOne","@refinedev/core"]),`
                const { data: ${P(r.key,"Data")}, isLoading: ${P(r.key,"IsLoading")} } =
                useOne({
                    resource: "${r.resource.name}",
                    id: ${d(u,r.key,r.accessor,!1)} || "",
                    queryOptions: {
                        enabled: !!${u},
                    },
                    ${D(((w=r==null?void 0:r.resource)==null?void 0:w.identifier)??((H=r==null?void 0:r.resource)==null?void 0:H.name),c,"getOne")}
                });
            `}}).filter(Boolean),T=o(r=>{if(r.relation&&r.resource){let h=P(r.key,"Data"),F=P(r.key,"IsLoading");if(r.multiple){let w=d(u,r.key)+"?.length";return x`
                <div style={{ marginTop: "6px" }}>
                    <h5>${k({resource:e,field:r,i18n:p,noQuotes:!0})}</h5>
                    <ul>
                    {${F} && ${w} ? <>Loading...</> : (
                        <>
                        ${(()=>{var H;if(r.relationInfer)if((H=r.relationInfer)!=null&&H.accessor){if(Array.isArray(r.relationInfer.accessor))return"";{let L=P(r.key),Q=d(L,void 0,r.relationInfer.accessor);return x`
                                            {record?.${r.key}?.length ? ${h}?.data?.map((${L}: any) => <li key={${Q}}>{${Q}}</li>) : <></>}
                                        `}}else return"";else return""})()}
                        </>
                    )}
                    </ul>
                </div>
                `}return r.fieldable?x`
                <div style={{ marginTop: "6px" }}>
                    <h5>${k({resource:e,field:r,i18n:p,noQuotes:!0})}</h5>
                    <div>{${d(u,r.key,r.accessor)}}</div>
                </div>
                `:x`
                <div style={{ marginTop: "6px" }}>
                    <h5>${k({resource:e,field:r,i18n:p,noQuotes:!0})}</h5>
                    <div>
                    {${F} ? <>Loading...</> : (
                        <>
                        ${(()=>{var w;return r.relationInfer?(w=r.relationInfer)!=null&&w.accessor?Array.isArray(r.relationInfer.accessor)?`{${d(`${h}?.data`,void 0,r.relationInfer.accessor,' + " " + ')}}`:`{${h}?.data?.${r.relationInfer.accessor}}`:`{${h}?.data}`:`{${h}?.data?.id}`})()}
                        </>
                    )}
                    </div>
                </div>
                `}},"renderRelationFields"),C=o(r=>{if(r.type==="text"){if(r.multiple){let h=d("item",void 0,r.accessor);return x`
                <div style={{ marginTop: "6px" }}>
                    <h5>${k({resource:e,field:r,i18n:p,noQuotes:!0})}</h5>
                    <ul>
                        {${d(u,r.key)}?.map((item: any) => (
                            <li key={${h}}>
                                {${h}}
                            </li>
                        ))}
                    </ul>
                </div>
            `}return x`
                <div style={{ marginTop: "6px" }}>
                    <h5>${k({resource:e,field:r,i18n:p,noQuotes:!0})}</h5>
                    <div>
                    {${d(u,r.key,r.accessor)}}
                    </div>
                </div>
            `}},"textFields"),$=o(r=>{if(r.type==="image"){if(r.multiple){let h=d("item",void 0,r.accessor);return x`
                <div style={{ marginTop: "6px" }}>
                    <h5>${k({resource:e,field:r,i18n:p,noQuotes:!0})}</h5>
                    <ul>
                    {${d(u,r.key)}?.map((item: any) => (
                        <li key={${h}}>
                            <img style={{ maxWidth: 200 }} src={${h}}/>
                        </li>
                    ))}
                    </ul>
                </div>
            `}return x`
                <div style={{ marginTop: "6px" }}>
                    <h5>${k({resource:e,field:r,i18n:p,noQuotes:!0})}</h5>
                    <img src={{ maxWidth: 200 }} src={${d(u,r.key,r.accessor," + ")}} />
                </div>
                `}},"imageFields"),A=o(r=>{if(r.type==="email"){if(r.multiple){let h=d("item",void 0,r.accessor);return x`
                <div style={{ marginTop: "6px" }}>
                    <h5>${k({resource:e,field:r,i18n:p,noQuotes:!0})}</h5>
                    <ul>
                        {${d(u,r.key)}?.map((item: any) => (
                            <li key={${h}}>
                                <a href={"mailto:" + ${h}}>{${h}}</a>
                            </li>
                        ))}
                    </ul>
                </div>
            `}return x`
                <div style={{ marginTop: "6px" }}>
                    <h5>${k({resource:e,field:r,i18n:p,noQuotes:!0})}</h5>
                    <div>
                        <a href={"mailto:" + ${d(u,r.key,r.accessor," + ")}}>
                            {${d(u,r.key,r.accessor," + ")}}
                        </a>
                    </div>
                </div>
            `}},"emailFields"),l=o(r=>{if(r.type==="url"){if(r.multiple){let h=d("item",void 0,r.accessor);return x`
                <div style={{ marginTop: "6px" }}>
                    <h5>${k({resource:e,field:r,i18n:p,noQuotes:!0})}</h5>
                    <ul>
                        {${d(u,r.key)}?.map((item: any) => (
                            <li key={${h}}>
                                <a href={${h}}>{${h}}</a>
                            </li>
                        ))}
                    </ul>
                </div>
            `}return x`
            <div style={{ marginTop: "6px" }}>
                <h5>${k({resource:e,field:r,i18n:p,noQuotes:!0})}</h5>
                <div>
                    <a href={${d(u,r.key,r.accessor," + ")}} target="_blank" rel="noreferrer">
                        {${d(u,r.key,r.accessor," + ")}}
                    </a>
                </div>
            </div>
            `}},"urlFields"),m=o(r=>{if(r.type==="boolean"){if(r.multiple){let h=d("item",void 0,r.accessor);return x`
                <div style={{ marginTop: "6px" }}>
                    <h5>${k({resource:e,field:r,i18n:p,noQuotes:!0})}</h5>
                    <ul>
                        {${d(u,r.key)}?.map((item: any) => (
                            <li key={${h}}>
                                {${h} ? "Yes" : "No"}
                            </li>
                        ))}
                    </ul>
                </div>
            `}return x`
            <div style={{ marginTop: "6px" }}>
                <h5>${k({resource:e,field:r,i18n:p,noQuotes:!0})}</h5>
                <div>
                {${d(u,r.key,r.accessor," && ")} ? "Yes" : "No"}
                </div>
            </div>
            `}},"booleanFields"),i=o(r=>{if(r.type==="date"){if(r.multiple){let h=d("item",void 0,r.accessor);return x`
                <div style={{ marginTop: "6px" }}>
                    <h5>${k({resource:e,field:r,i18n:p,noQuotes:!0})}</h5>
                    <ul>
                    {${d(u,r.key)}?.map((item: any) => (
                        <li key={${h}}>
                            {(new Date(${h})).toLocaleString(undefined, { timeZone: "UTC" })}
                        </li>
                    ))}
                    </ul>
                </div>
            `}return x`
            <div style={{ marginTop: "6px" }}>
                <h5>${k({resource:e,field:r,i18n:p,noQuotes:!0})}</h5>
                <div>
                    {(new Date(${d(u,r.key,r.accessor," + ' ' + ")})).toLocaleString(undefined, { timeZone: "UTC" })}
                </div>
            </div>
            `}},"dateFields"),b=o(r=>{if(r.type==="richtext")return x`
            <div style={{ marginTop: "6px" }}>
                <h5>${k({resource:e,field:r,i18n:p,noQuotes:!0})}</h5>
                <p>
                {${d(u,r.key,r.accessor,' + " " + ')}}
                </p>
            </div>
            `},"richtextFields"),N=o(r=>{if(r.type==="number"){if(r.multiple){let h=d("item",void 0,r.accessor);return x`
                <div style={{ marginTop: "6px" }}>
                    <h5>${k({resource:e,field:r,i18n:p,noQuotes:!0})}</h5>
                    <ul>
                        {${d(u,r.key)}?.map((item: any) => (
                        <li key={${h}}>
                            {${h}}
                        </li>
                        ))}
                    </ul>
                </div>
                `}return x`
                <div style={{ marginTop: "6px" }}>
                    <h5>${k({resource:e,field:r,i18n:p,noQuotes:!0})}</h5>
                    <div>
                    {${d(u,r.key,r.accessor,' + " " + ')} ?? ""}
                    </div>
                </div>
            `}},"numberFields"),S=o(r=>{if(r)return x`
                ${r}
            `},"wrapper"),B=t.map(r=>{switch(r==null?void 0:r.type){case"text":return S(C(r));case"number":return S(N(r));case"richtext":return S(b(r));case"email":return S(A(r));case"image":return S($(r));case"date":return S(i(r));case"boolean":return S(m(r));case"url":return S(l(r));case"relation":return S(T(r));default:return}}),{canEdit:_,list:R}=e??{},M=!!R;ce(f);let n=p&&"const translate = useTranslate();";return x`
    ${ie(f)}
    
    export const ${y}: React.FC<IResourceComponentsProps> = () => {
        ${n}
        const { edit, list } = useNavigation();
        ${a?"":"const { id } = useResource();"}
        const { queryResult } = useShow(${a?`{ 
                    resource: "${e.name}", 
                    id: ${Se(s)},
                    ${D((e==null?void 0:e.identifier)??(e==null?void 0:e.name),c,"getOne")}
                }`:D((e==null?void 0:e.identifier)??(e==null?void 0:e.name),c,"getOne")?`{ ${D((e==null?void 0:e.identifier)??(e==null?void 0:e.name),c,"getOne")} }`:""});
        const { data, isLoading } = queryResult;

        const ${u} = data?.data;
    
        ${v}

        return (
            <div style={{ padding: "16px" }}>
            <div style={{ display: "flex", alignItems: "center", justifyContent: "space-between" }}>
                <h1>${ee({resource:e,action:"show",i18n:p})}</h1>
                <div style={{ display: "flex", gap: "8px" }}>
                ${M?x`<button onClick={() => list("${e.name}")}>${ee({resource:e,action:"list",i18n:p})}</button>`:""}
                ${_?x`<button onClick={() => edit("${e.name}", ${a?`"${s}"`:"id ?? ''"})}>${ne({action:"edit",i18n:p,noQuotes:!0})}</button>`:""}
                </div>
            </div>
                <div>
                    ${B.join("")}
                </div>
            </div>
        );
    };
    `},"renderer"),Ve=de({type:"show",additionalScope:[],codeViewerComponent:le,loadingComponent:ge,errorComponent:fe,renderer:rt});var _t=require("@refinedev/react-table"),Wt=require("@tanstack/react-table");var ye=o(e=>Array.isArray(e.accessor)||e.multiple?`accessorKey: "${e.key}"`:e.accessor?`accessorKey: "${q(e.key,void 0,e.accessor)}"`:`accessorKey: "${e.key}"`,"getAccessorKey"),nt=o(({resource:e,fields:t,meta:c,isCustomPage:a,i18n:s})=>{let p=oe(e.label??e.name,"list"),y="tableData?.data",u=[["React","react",!0],["IResourceComponentsProps","@refinedev/core"],["useNavigation","@refinedev/core"],["useTable","@refinedev/react-table"],["ColumnDef","@tanstack/react-table"],["flexRender","@tanstack/react-table"]];s&&u.push(["useTranslate","@refinedev/core"]);let f=t.filter(n=>(n==null?void 0:n.relation)&&!(n!=null&&n.fieldable)&&(n==null?void 0:n.resource)),I=f.filter(Boolean).map(n=>{var r,h;if(n!=null&&n.relation&&!n.fieldable&&n.resource){u.push(["GetManyResponse","@refinedev/core"]),u.push(["useMany","@refinedev/core"]);let F="";return n.multiple?F=`[].concat(...(${y}?.map((item) => ${d("item",n.key,n.accessor,!1)}) ?? []))`:F=`${y}?.map((item) => ${d("item",n.key,n.accessor,!1)}) ?? []`,`
                const { data: ${P(n.key,"Data")} } =
                useMany({
                    resource: "${n.resource.name}",
                    ids: ${F},
                    queryOptions: {
                        enabled: !!${y},
                    },
                    ${D(((r=n==null?void 0:n.resource)==null?void 0:r.identifier)??((h=n==null?void 0:n.resource)==null?void 0:h.name),c,"getMany")}
                });
                `}}).filter(Boolean),v=f==null?void 0:f.map(n=>{if(n&&n.resource)return P(n.key,"Data")}).filter(Boolean),T=o(n=>{var r;if(n.relation&&n.resource){let h=`${P(n.key,"Data")}?.data`;if(Array.isArray(n.accessor))return;let F=`id: "${n.key}"`,w=`header: ${k({resource:e,field:n,i18n:s,noBraces:!0})}`,H=ye(n),L="";if(n.multiple){let Q="item";n!=null&&n.relationInfer&&(Q=d("item",void 0,n.relationInfer.accessor)),L=`cell: function render({ getValue, table }) {
                    const meta = table.options.meta as {
                        ${P(n.key,"Data")}: GetManyResponse;
                    };

                    ${n!=null&&n.accessor?"try {":""}

                    const ${P(n.key,"")} = getValue<any[]>()?.map((item) => {
                        return meta.${P(n.key,"Data")}?.data?.find(
                            (resourceItems) => resourceItems.id === ${d("item",void 0,n.accessor)}
                        );
                    })


                    return (
                        <ul>
                            {${P(n.key,"")}?.map((item, index) => (
                                <li key={index}>
                                    {${Q}}
                                </li>
                            ))}
                        </ul>
                    )
                    ${n!=null&&n.accessor?" } catch (error) { return null; }":""}
                }
            `}else n!=null&&n.relationInfer?L=`cell: function render({ getValue, table }) {
                        const meta = table.options.meta as {
                            ${P(n.key,"Data")}: GetManyResponse;
                        };

                        ${n!=null&&n.accessor?"try {":""}

                        const ${P(n.key,"")} = meta.${h}?.find(
                            (item) => item.id == getValue<any>(),
                        );

                        return ${d(P(n.key),void 0,(r=n==null?void 0:n.relationInfer)==null?void 0:r.accessor)} ?? "Loading...";

                        ${n!=null&&n.accessor?" } catch (error) { return null; }":""}
                    },`:L="";return`
                {
                    ${F},
                    ${w},
                    ${H},
                    ${L}
                }
            `}},"renderRelationFields"),C=o(n=>{if(n.type==="image"){let r=`id: "${n.key}"`,h=ye(n),F=`header: ${k({resource:e,field:n,i18n:s,noBraces:!0})}`,w=x`
                cell: function render({ getValue }) {
                    ${n!=null&&n.accessor?"try {":""}
                        return <img style={{ maxWidth: "100px" }} src={${d("getValue<any>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0," + ")}} />
                    ${n!=null&&n.accessor?" } catch (error) { return null; }":""}
                }
            `;if(n.multiple){let H=d("item",void 0,n.accessor," + ");w=`
                    cell: function render({ getValue }) {
                        ${n!=null&&n.accessor?"try {":""}
                            return (
                                <ul>
                                    {getValue<any[]>()?.map((item, index) => (
                                        <li key={index}><img src={${H}} style={{ height: "50px", maxWidth: "100px" }} /></li>
                                    ))}
                                </ul>
                            )
                        ${n!=null&&n.accessor?" } catch (error) { return null; }":""}
                    }
                `}return`
                {
                    ${r},
                    ${h},
                    ${F},
                    ${w}
                }
            `}},"imageFields"),$=o(n=>{if(n.type==="email"){let r=`id: "${n.key}"`,h=ye(n),F=`header: ${k({resource:e,field:n,i18n:s,noBraces:!0})}`,w=x`
                cell: function render({ getValue }) {
                    return <a href={"mailto:" + ${d("getValue<any>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0,' + " " + ')}}>{${d("getValue<any>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0,' + " " + ')}}</a>
                }
            `;return n.multiple&&(w=`
                    cell: function render({ getValue }) {
                        return (
                            <ul>
                                {getValue<any[]>()?.map((item, index) => (
                                    <li key={index}>
                                        {${d("item",void 0,n.accessor," + ")}}
                                    </li>
                                ))}
                            </ul>
                        )
                    }
                `),`
                {
                    ${r},
                    ${h},
                    ${F},
                    ${w}
                }
            `}},"emailFields"),A=o(n=>{if(n.type==="url"){let r=`id: "${n.key}"`,h=ye(n),F=`header: ${k({resource:e,field:n,i18n:s,noBraces:!0})}`,w=x`
                cell: function render({ getValue }) {
                    return <a href={${d("getValue<any>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0," + ")}}>{${d("getValue<any>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0," + ")}}</a>
                }
            `;return n.multiple&&(w=`
                    cell: function render({ getValue }) {
                        return (
                            <ul>
                                {getValue<any[]>()?.map((item, index) => (
                                    <li key={index}>
                                        {${d("item",void 0,n.accessor," + ")}}
                                    </li>
                                ))}
                            </ul>
                        )
                    }
                `),`
                {
                    ${r},
                    ${h},
                    ${F},
                    ${w}
                }
            `}},"urlFields"),l=o(n=>{if((n==null?void 0:n.type)==="boolean"){let r=`id: "${n.key}"`,h=ye(n),F=`header: ${k({resource:e,field:n,i18n:s,noBraces:!0})}`,w=x`
                cell: function render({ getValue }) {
                    return ${d("getValue<any>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0," + ")} ? "yes" : "no"
                }
            `;return n.multiple&&(w=`
                    cell: function render({ getValue }) {
                        return (
                            <ul>
                                {getValue<any[]>()?.map((item, index) => (
                                    <li key={index}>
                                        {${d("item",void 0,n.accessor," + ")} ? "yes" : "no"}
                                    </li>
                                ))}
                            </ul>
                        );
                    }
                `),`
                {
                    ${r},
                    ${h},
                    ${F},
                    ${w}
                }
            `}},"booleanFields"),m=o(n=>{if(n.type==="date"){let r=`id: "${n.key}"`,h=ye(n),F=`header: ${k({resource:e,field:n,i18n:s,noBraces:!0})}`,w=x`
                cell: function render({ getValue }) {
                    return (new Date(${d("getValue<any>()",void 0,Array.isArray(n.accessor)?n.accessor:void 0,' + " " + ')})).toLocaleString(undefined, { timeZone: "UTC" })
                }
            `;return n.multiple&&(w=`
                    cell: function render({ getValue }) {
                        return (
                            <ul>
                                {getValue<any[]>()?.map((item, index) => (
                                    <li key={index}>
                                    {(new Date(${d("item",void 0,n.accessor," + ")})).toLocaleString(undefined, { timeZone: "UTC" })}
                                    </li>
                                ))}
                            </ul>
                        )
                    }
                `),`
                {
                    ${r},
                    ${h},
                    ${F},
                    ${w}
                }
            `}},"dateFields"),i=o(n=>{if(n&&(n.type==="text"||n.type==="number"||n.type==="richtext")){let r=`id: "${n.key}"`,h=ye(n),F=`header: ${k({resource:e,field:n,i18n:s,noBraces:!0})}`,w="";return n.multiple&&(w=`
                    cell: function render({ getValue }) {
                        return (
                            <ul>
                                {getValue<any[]>()?.map((item, index) => (
                                    <li key={index}>
                                        {${d("item",void 0,n.accessor,' + " " + ')}}
                                    </li>
                                ))}
                            </ul>
                        )
                    }
                `),!n.multiple&&Array.isArray(n.accessor)&&(w=`
                    cell: function render({ getValue }) {
                        return (
                            <>{${d("getValue<any>()",n.key,n.accessor)}}</>
                        );
                    }
                `),`
                {
                    ${r},
                    ${h},
                    ${F},
                    ${w}
                }
            `}},"basicFields"),{canEdit:b,canShow:N,canCreate:S}=e??{},_=b||N?x`
    {
        id: "actions",
        accessorKey: "id",
        header: ${s?'translate("table.actions")':'"Actions"'},
        cell: function render({ getValue }) {
            return (
                <div
                    style={{
                        display: "flex",
                        flexDirection: "row",
                        flexWrap: "wrap",
                        gap: "4px",
                    }}
                >
                ${N?x`
                    <button
                        onClick={() => {
                            show("${e.name}", getValue() as string);
                        }}
                    >
                        ${ne({action:"show",i18n:s,noQuotes:!0})}
                    </button>
                    `:""}
                    ${b?x`
                            <button
                            onClick={() => {
                                edit("${e.name}", getValue() as string);
                            }}
                        >
                            ${ne({action:"edit",i18n:s,noQuotes:!0})}
                        </button>
                    `:""}
                </div>
            );
        },
    },
        `:"",R=t.map(n=>{switch(n==null?void 0:n.type){case"text":case"number":case"richtext":return i(n);case"email":return $(n);case"image":return C(n);case"date":return m(n);case"boolean":return l(n);case"url":return A(n);case"relation":return T(n);default:return}});ce(u);let M=s&&"const translate = useTranslate();";return x`
    ${ie(u)}
    
    export const ${p}: React.FC<IResourceComponentsProps> = () => {
        ${M}
        const columns = React.useMemo<ColumnDef<any>[]>(() => [
            ${[...R,_].filter(Boolean).join(",")}
        ], [${s?"translate":""}]);

        ${b||N?x`
        const { ${b?"edit,":""} ${N?"show,":""} ${S?"create,":""} } = useNavigation();
        `:""}

        const {
            getHeaderGroups,
            getRowModel,
            setOptions,
            refineCore: {
                tableQueryResult: { data: tableData },
            },
            getState,
            setPageIndex,
            getCanPreviousPage,
            getPageCount,
            getCanNextPage,
            nextPage,
            previousPage,
            setPageSize,
            getColumn,
        } = useTable({
            columns,
            ${a?`
            refineCoreProps: {
                resource: "${e.name}",
                ${D((e==null?void 0:e.identifier)??(e==null?void 0:e.name),c,"getList")}
            }
            `:D((e==null?void 0:e.identifier)??(e==null?void 0:e.name),c,"getList")?`refineCoreProps: { ${D((e==null?void 0:e.identifier)??(e==null?void 0:e.name),c,"getList")} },`:""}
            
        });

        ${I}

        setOptions((prev) => ({
            ...prev,
            meta: {
                ...prev.meta,
                ${v.join(", ")}
            },
        }));

        return (
            <div style={{ padding: "16px" }}>
            <div style={{ display: "flex", alignItems: "center", justifyContent: "space-between" }}>
                <h1>${ee({resource:e,action:"list",i18n:s})}</h1>
                ${S?x`<button onClick={() => create("${e.name}")}>${ne({action:"create",i18n:s,noQuotes:!0})}</button>`:""}
            </div>
                <div style={{ maxWidth: "100%", overflowY: "scroll" }}>
                    <table>
                        <thead>
                            {getHeaderGroups().map((headerGroup) => (
                                <tr key={headerGroup.id}>
                                    {headerGroup.headers.map((header) => (
                                        <th key={header.id}>
                                            {!header.isPlaceholder && (
                                                flexRender(
                                                    header.column.columnDef
                                                        .header,
                                                    header.getContext(),
                                                )
                                            )}
                                        </th>
                                    ))}
                                </tr>
                            ))}
                        </thead>
                        <tbody>
                            {getRowModel().rows.map((row) => (
                                <tr key={row.id}>
                                    {row.getVisibleCells().map((cell) => (
                                        <td key={cell.id}>
                                            {flexRender(
                                                cell.column.columnDef.cell,
                                                cell.getContext(),
                                            )}
                                        </td>
                                    ))}
                                </tr>
                            ))}
                        </tbody>
                    </table>
                </div>
                <div style={{ marginTop: "12px" }}>
                    <button
                        onClick={() => setPageIndex(0)}
                        disabled={!getCanPreviousPage()}
                    >
                        {"<<"}
                    </button>
                    <button
                        onClick={() => previousPage()}
                        disabled={!getCanPreviousPage()}
                    >
                        {"<"}
                    </button>
                    <button onClick={() => nextPage()} disabled={!getCanNextPage()}>
                        {">"}
                    </button>
                    <button
                        onClick={() => setPageIndex(getPageCount() - 1)}
                        disabled={!getCanNextPage()}
                    >
                        {">>"}
                    </button>
                    <span>
                        <strong>
                            {" "} {getState().pagination.pageIndex + 1} / {" "} {getPageCount()} {" "}
                        </strong>
                    </span>
                    <span>
                        | ${s?'{translate("pagination.go")}':"Go to Page"}:{" "}
                        <input
                            type="number"
                            defaultValue={getState().pagination.pageIndex + 1}
                            onChange={(e) => {
                                const page = e.target.value
                                    ? Number(e.target.value) - 1
                                    : 0;
                                setPageIndex(page);
                            }}
                        />
                    </span>{" "}
                    <select
                        value={getState().pagination.pageSize}
                        onChange={(e) => {
                            setPageSize(Number(e.target.value));
                        }}
                    >
                        {[10, 20, 30, 40, 50].map((pageSize) => (
                            <option key={pageSize} value={pageSize}>
                                ${s?'{translate("pagination.show")}':"Show"} {pageSize}
                            </option>
                        ))}
                    </select>
                </div>
            </div>   
        );
    };
    `},"renderer"),Ee=de({type:"list",additionalScope:[["@refinedev/react-table","RefineReactTable",{useTable:_t.useTable}],["@tanstack/react-table","TanstackReactTable",{flexRender:Wt.flexRender}]],codeViewerComponent:le,loadingComponent:ge,errorComponent:fe,renderer:nt});var qt=require("@refinedev/react-hook-form");var ot=o(({resource:e,fields:t,meta:c,isCustomPage:a,i18n:s})=>{let p=oe(e.label??e.name,"create"),y=[["React","react",!0],["useNavigation","@refinedev/core"],["IResourceComponentsProps","@refinedev/core"],["useForm","@refinedev/react-hook-form"]];s&&y.push(["useTranslate","@refinedev/core"]);let f=t.filter(l=>(l==null?void 0:l.relation)&&!(l!=null&&l.fieldable)&&(l==null?void 0:l.resource)).filter(Boolean).map(l=>{var m,i;if(l!=null&&l.relation&&!l.fieldable&&l.resource)return y.push(["useSelect","@refinedev/core"]),`
                const { options: ${P(l.key,"Options")} } =
                useSelect({
                    resource: "${l.resource.name}",
                    ${we(l)}
                    ${D(((m=l==null?void 0:l.resource)==null?void 0:m.identifier)??((i=l==null?void 0:l.resource)==null?void 0:i.name),c,"getList")}
                });
            `}).filter(Boolean),I=o(l=>{if(l.relation&&l.resource){y.push(["useSelect","@refinedev/core"]);let m=P(l.key,"Options");return x`
            <label>
                <span style={{ marginRight: "8px" }}>
                    ${k({resource:e,field:l,i18n:s,noQuotes:!0})}
                </span>
                <select
                    placeholder="Select ${W(l.resource.name)}"
                    {...register("${q(l.key,void 0,l.accessor)}", {
                        required: ${l.multiple?"false":'"This field is required"'},
                    })}
                >
                    {${m}?.map((option) => (
                        <option value={option.value} key={option.value}>
                            {option.label}
                        </option>
                    ))}
                </select>
                <span style={{ color: "red" }}>
                    {${d("(errors as any)",l.key,l.accessor,!1)}?.message as string}
                </span>
            </label>
            `}},"renderRelationFields"),v=o(l=>{if(l.type==="text"||l.type==="url"||l.type==="email"||l.type==="number"||l.type==="date"||l.type==="richtext"){if(Fe(l.key)||l.multiple)return;let m=l.type==="richtext"?"textarea":"input";return x`
                <label>
                    <span style={{ marginRight: "8px" }}>
                        ${k({resource:e,field:l,i18n:s,noQuotes:!0})}
                    </span>
                    <${m}
                        ${l.type!=="date"&&l.type!=="richtext"?`type="${l.type}"`:""}
                        ${m==="textarea"?`
                            rows={5}
                            cols={33}
                            style={{ verticalAlign: "top" }}
                        `:""}
                        {...register("${q(l.key,void 0,l.accessor)}", {
                            required: "This field is required",
                            ${l.type==="number"?"valueAsNumber: true,":""}
                        })}

                    />
                    <span style={{ color: "red" }}>
                        {${d("(errors as any)",l.key,l.accessor,!1)}?.message as string}
                    </span>
                </label>
            `}},"basicInputFields"),T=o(l=>{if(l.type==="boolean")return l.multiple?void 0:x`
                <label>
                    <span style={{ marginRight: "8px" }}>
                        ${k({resource:e,field:l,i18n:s,noQuotes:!0})}
                    </span>
                    <input
                        type="checkbox"
                        {...register("${q(l.key,void 0,l.accessor)}", {
                            required: "This field is required",
                        })}
                    />
                    <span style={{ color: "red" }}>
                        {${d("errors",l.key,l.accessor,!1)}?.message as string}
                    </span>
                </label>
            `},"booleanFields"),C=t.map(l=>{switch(l==null?void 0:l.type){case"text":case"number":case"email":case"url":case"richtext":case"date":return v(l);case"boolean":return T(l);case"relation":return I(l);default:return}}),$=!!e.list;ce(y);let A=s&&"const translate = useTranslate();";return x`
    ${ie(y)}
    
    export const ${p}: React.FC<IResourceComponentsProps> = () => {
        ${A}
        ${$?`
        const { list } = useNavigation();
        `:""}
        const {
            refineCore: { onFinish, formLoading },
            register,
            handleSubmit,
            resetField,
            formState: { errors },
        } = useForm(
            ${a?`
            { 
                refineCoreProps: {
                    resource: "${e.name}",
                    action: "create",
                    ${D(e.identifier??e.name,c,"getOne")}
                }
            }`:D(e.identifier??e.name,c,"getOne")?`{
                        refineCoreProps: { ${D(e.identifier??e.name,c,"getOne")} }
                        }`:""}
        );
    
        ${f}

        return (
            <div style={{ padding: "16px" }}>
                <div style={{ display: "flex", justifyContent: ${$?'"space-between"':'"flex-start"'} }}>
                    <h1>${ee({resource:e,action:"create",i18n:s})}</h1>
                    ${$?x`
                            <div>
                        <button
                                onClick={() => {
                                    list("${e.name}");
                                }}
                        >
                            ${ee({resource:e,action:"list",i18n:s})}
                        </button>
                        </div>
                    `:""}
                </div>
                <form onSubmit={handleSubmit(onFinish)}>
                    <div style={{ display: "flex", flexDirection: "column", gap: "8px" }}>
                        ${C.join("")}
                        <div>
                            <input type="submit" value=${ne({action:"save",i18n:s})} />
                        </div>
                    </div>
                </form>
            </div>
        );
    };
    `},"renderer"),Me=de({type:"create",additionalScope:[["@refinedev/react-hook-form","RefineReactHookForm",{useForm:qt.useForm}]],codeViewerComponent:le,loadingComponent:ge,errorComponent:fe,renderer:ot});var Gt=require("@refinedev/react-hook-form");var st=o(({resource:e,fields:t,meta:c,isCustomPage:a,id:s,i18n:p})=>{let y=oe(e.label??e.name,"edit"),u=P(e.label??e.name,"Data"),f=[["React","react",!0],["useNavigation","@refinedev/core"],["IResourceComponentsProps","@refinedev/core"],["useForm","@refinedev/react-hook-form"]];p&&f.push(["useTranslate","@refinedev/core"]);let v=t.filter(i=>(i==null?void 0:i.relation)&&!(i!=null&&i.fieldable)&&(i==null?void 0:i.resource)).filter(Boolean).map(i=>{var b,N;if(i!=null&&i.relation&&!i.fieldable&&i.resource){f.push(["useSelect","@refinedev/core"]);let S=d(u,i.key,i.accessor,!1);return i.multiple&&i.accessor&&(S=`${d(u,i.key)}?.map((item: any) => ${d("item",void 0,i.accessor)})`),`
                const { options: ${P(i.key,"Options")} } =
                useSelect({
                    resource: "${i.resource.name}",
                    defaultValue: ${S},
                    ${we(i)}
                    ${D(((b=i==null?void 0:i.resource)==null?void 0:b.identifier)??((N=i==null?void 0:i.resource)==null?void 0:N.name),c,"getList")}
                });

                React.useEffect(() => {
                    setValue("${q(i.key,void 0,i.accessor)}", ${u}?.${i.key}?.id);
                }, [${P(i.key,"Options")}]);
            `}}).filter(Boolean),T=o(i=>{if(i.relation&&i.resource){f.push(["useSelect","@refinedev/core"]);let b=P(i.key,"Options");return x`
            <label>
                <span style={{ marginRight: "8px" }}>${k({resource:e,field:i,i18n:p,noQuotes:!0})}</span>
                <select
                    placeholder="Select ${W(i.resource.name)}"
                    {...register("${q(i.key,void 0,i.accessor)}", {
                        required: ${i.multiple?"false":'"This field is required"'},
                    })}
                >
                    {${b}?.map((option) => (
                        <option value={option.value} key={option.value}>
                            {option.label}
                        </option>
                    ))}
                </select>
                <span style={{ color: "red" }}>
                    {${d("(errors as any)",i.key,i.accessor,!1)}?.message as string}
                </span>
            </label>
            `}},"renderRelationFields"),C=o(i=>{if(i.type==="text"||i.type==="url"||i.type==="email"||i.type==="number"||i.type==="date"||i.type==="richtext"){if(i.multiple){let N=q(i.key,"${index}",i.accessor),S=d(d("(errors as any)",i.key,void 0,!1)+"?.[index]",void 0,i.accessor);return`
                    <>
                        {${d(u,i.key)}?.map((item: any, index: number) => (
                            <label key={index}>
                                <span style={{ marginRight: "8px" }}>
                                    ${k({resource:e,field:i,i18n:p,noQuotes:!0})}
                                </span>
                                <input ${i.type!=="richtext"?`
                                type="${i.type!=="date"?i.type:"text"}"
                                `:""} {...register(\`${N}\`, { required: "This field is required", ${i.type==="number"?"valueAsNumber: true,":""} })} />
                                <span style={{color: "red"}}>
                                {${d(S,"message")} as string}
                                </span>
                            </label>
                        ))}
                    </>
                `}let b=i.type==="richtext"?"textarea":"input";return x`
            <label>
                <span style={{ marginRight: "8px" }}>${k({resource:e,field:i,i18n:p,noQuotes:!0})}</span>
                <${b}
                ${Fe(i.key)?"disabled":""}
                ${i.type!=="date"&&i.type!=="richtext"?`type="${i.type}"`:""}
                ${b==="textarea"?`
                    rows={5}
                    cols={33}
                    style={{ verticalAlign: "top" }}
                `:""}
                {...register("${q(i.key,void 0,i.accessor)}", {
                    required: "This field is required",
                    ${i.type==="number"?"valueAsNumber: true,":""}
                })}
                />
                <span style={{ color: "red" }}>
                {${d("(errors as any)",i.key,i.accessor,!1)}?.message as string}
                </span>
            </label>
            `}},"basicInputFields"),$=o(i=>{if(i.type==="boolean"){if(i.multiple){let b=q(i.key,void 0,i.accessor),N=d("(errors as any)",i.key,void 0,!1)+"?.[index]";return`
                    <>
                        {${d(u,i.key)}?.map((item: any, index: number) => (
                            <label key={index}>
                            <span style={{ marginRight: "8px" }}>
                            ${k({resource:e,field:i,i18n:p,noQuotes:!0})}
                            </span>
                            <input
                                type="checkbox"
                                {...register(\`${b}.\${index}\`, {
                                    required: "This field is required",
                                })}
                            />
                            <span style={{ color: "red" }}>
                                {${N}?.message as string}
                            </span>
                        ))}
                    </>
                `}return x`
            <label>
                <span style={{ marginRight: "8px" }}>${k({resource:e,field:i,i18n:p,noQuotes:!0})}</span>
                <input type="checkbox" {...register("${q(i.key,void 0,i.accessor)}", {
                    required: "This field is required",
                })} />
                <span style={{ color: "red" }}>
                    {${d("errors",i.key,i.accessor,!1)}?.message as string}
                </span>
            </label>
            `}},"booleanFields"),A=t.map(i=>{switch(i==null?void 0:i.type){case"text":case"number":case"email":case"url":case"richtext":case"date":return C(i);case"boolean":return $(i);case"relation":return T(i);default:return}}),l=!!e.list;ce(f);let m=p&&"const translate = useTranslate();";return x`
    ${ie(f)}
    
    export const ${y}: React.FC<IResourceComponentsProps> = () => {
        ${m}
        ${l?`
        const { list } = useNavigation();
        `:""}
        const {
            refineCore: { onFinish, formLoading, queryResult },
            register,
            handleSubmit,
            setValue,
            formState: { errors },
        } = useForm(
            ${a?`
            { 
                refineCoreProps: {
                    resource: "${e.name}",
                    id: ${Se(s)},
                    action: "edit",
                    ${D((e==null?void 0:e.identifier)??(e==null?void 0:e.name),c,"getOne")}
                }
            }`:D((e==null?void 0:e.identifier)??(e==null?void 0:e.name),c,"getOne")?`{ refineCoreProps: {
                        ${D((e==null?void 0:e.identifier)??(e==null?void 0:e.name),c,"getOne")}
                    } }`:""}
        );
    
        const ${u} = queryResult?.data?.data;
    
        ${v}

        return (
            <div style={{ padding: "16px" }}>
                <div style={{ display: "flex", justifyContent: ${l?'"space-between"':'"flex-start"'} }}>
                    <h1>${ee({resource:e,action:"edit",i18n:p})}</h1>
                    ${l?x`
                    <div>
                        <button
                                onClick={() => {
                                    list("${e.name}");
                                }}
                        >
                            ${ee({resource:e,action:"list",i18n:p})}
                        </button>
                    </div>
                    `:""}
                </div>
                <form onSubmit={handleSubmit(onFinish)}>
                    <div style={{ display: "flex", flexDirection: "column", gap: "8px" }}>
                        ${A.join("")}
                        <div>
                            <input type="submit" value=${ne({action:"save",i18n:p})} />
                        </div>
                    </div>
                </form>
            </div>
        );
    };
    `},"renderer"),Oe=de({type:"edit",additionalScope:[["@refinedev/react-hook-form","RefineReactHookForm",{useForm:Gt.useForm}]],codeViewerComponent:le,loadingComponent:ge,errorComponent:fe,renderer:st});var kr=o(({action:e,id:t,...c})=>{let{action:a,id:s}=(0,Zt.useResource)();switch(e??a){case"show":return Ae.default.createElement(Ve,{...c,id:t??s});case"create":return Ae.default.createElement(Me,{...c,id:t??s});case"edit":return Ae.default.createElement(Oe,{...c,id:t??s});default:return Ae.default.createElement(Ee,{...c,id:t??s})}},"HeadlessInferencer");
